### PR TITLE
ARROW-10729: [Rust] [DataFusion] Add SQL support for JOIN using implicit syntax

### PR DESCRIPTION
**Pull request by andygrove**
Adds support for SQL joins using the implicit syntax where the "ON" conditions are part of the WHERE clause.

For example:

```sql
SELECT t1_id, t1_name, t2_name FROM t1, t2 WHERE t2_id = t1_id ORDER BY t1_id
```

### CODE REVIEW

**Comment by andygrove** [comment]
@jorgecarleitao This code is basically working but I think it has exposed a bug in the filter push down logic where it tries to push a filter down through the join where the expression is `left_column = right_column` so it is not valid to push this through the join because it depends on the output from the join.

```
---- equijoin_implicit_syntax stdout ----
thread 'equijoin_implicit_syntax' panicked at 'Creating physical plan for 'SELECT t1_id, t1_name, t2_name FROM t1, t2 WHERE t1_id = t2_id ORDER BY t1_id': Sort: #t1_id ASC NULLS FIRST
  Projection: #t1_id, #t1_name, #t2_name
    Join: t1_id = t2_id
      Filter: #t1_id Eq #t2_id
        TableScan: t1 projection=Some([0, 1])
      Filter: #t1_id Eq #t2_id
        TableScan: t2 projection=Some([0, 1]): ArrowError(InvalidArgumentError("Unable to get field named \"t2_id\". Valid fields: [\"t1_id\", \"t1_name\"]"))', datafusion/tests/sql.rs:1243:48
```

**Comment by jorgecarleitao** [comment]
Thanks @andygrove. 

I am a bit confused: why is there a filter on the plan? I thought that the logical plan of 

```
SELECT t1_id, t1_name, t2_name FROM t1, t2 WHERE t2_id = t1_id ORDER BY t1_id
```

would not contain a filter, as `=` in `WHERE t2_id = t1_id` represents "equal under hashing on a join" (that is already represented on the `Join: t1_id = t2_id` bit of the plan.

**Comment by andygrove** [comment]
With this syntax, the WHERE clause can contain both JOIN conditions and regular filter conditions. For example:

```sql
FROM
    customer,
    orders,
    lineitem
WHERE
    c_mktsegment = 'BUILDING'
    AND c_custkey = o_custkey
    AND l_orderkey = o_orderkey
    AND o_orderdate < date '1995-03-15'
    AND l_shipdate > date '1995-03-15'
```

I think there are two options for addressing this.

1. Have the SQL planner remove the join expressions from the filter expression
2. Modify the filter push down logic to prevent filters being pushed through joins when they reference both join inputs

I'm not sure which approach is best.

**Comment by jorgecarleitao** [comment]
Got it. I understand now.

I agree that there is a bug: the push down should not push down like it is doing now. 

I also think that the filter should not be there, as it is an extra filter equivalent to "1 = 1".

The former is definitely more important to fix than the second :) I will try to address it this weekend.

**Comment by andygrove** [comment]
@jorgecarleitao I made some improvements to remove any join expressions from the filter which helped for some cases but there is still one failing test related to filter push down. No rush to look at this but just wanted to update you.

**Comment by alamb** [comment]
> @jorgecarleitao This code is basically working but I think it has exposed a bug in the filter push down logic where it tries to push a filter down through the join where the expression is left_column = right_column so it is not valid to push this through the join because it depends on the output from the join.

I think another way to think about this is that "the expression refers to two relations and thus can only be evaluated when they have already been joined"

One way to represent this is to track which relation each `Column` reference is to (aka with multiple tables in a SQL query, a single column name is insufficient to identify what column is being referred to).

Until we add "to what what relation is this column" into the column reference itself, this kind of plan will be tricky to pull off

File: rust/datafusion/src/sql/planner.rs
Comment (alamb) [inline rust/datafusion/src/sql/planner.rs:332]: I am not sure how this code handles duplicated column names in the two inputs. TPCH helpfully prefixes id columns with the relation (e.g. `l_orderkey` and `o_orderkey`) but in many real world inputs both tables would probably have a column named `orderkey`)

**Review by alamb — APPROVED** [review]
I think this code looks like a good step in the right direction. I suggest adding an error if there are repeated column names in the two inputs so we don't end up with hard to debug issues later on. I think adding proper support for sch expressions could definitely be done as a follow on PR

My rationale is that "this PR adds features we don't yet have and are useable" -- even though of course more can always be done.

👍 from me

File: rust/datafusion/src/sql/planner.rs
Comment (andygrove) [reply rust/datafusion/src/sql/planner.rs:332]: Thanks @alamb. I will add the check for duplicate column names and merge this once the predicate push-down bug is resolved.

After this, I plan on working on table/relation aliases and qualified column names in queries [1].

https://issues.apache.org/jira/browse/ARROW-10732

Comment (andygrove) [reply rust/datafusion/src/sql/planner.rs:331]: On second thoughts, I have ignored the failing test and added a link to the bug [1].

I have also added the check for duplicate column names after a join.

[1] https://issues.apache.org/jira/browse/ARROW-10760

**Comment by andygrove** [comment]
I rebased after merging the join test and then re-enabled the test that was previously failing.

**Review by jorgecarleitao — APPROVED** [review]
Went through this and it looks great. Thanks a lot, @andygrove

**Review by alamb — APPROVED** [review]

**Pull request closed without merging**
